src/core/lib/iomgr/call_combiner.cc

```diff
@@ -39,10 +39,42 @@ static gpr_atm encode_cancel_state_error(grpc_error* error) {
   return static_cast<gpr_atm>(1) | (gpr_atm)error;
 }
 
+#ifdef GRPC_TSAN_ENABLED
+static void tsan_closure(void* user_data, grpc_error* error) {
+  grpc_call_combiner* call_combiner =
+      static_cast<grpc_call_combiner*>(user_data);
+  grpc_core::RefCountedPtr<grpc_call_combiner::TsanLock> lock;
+  if (!call_combiner->tsan_lock->taken.load(std::memory_order_relaxed)) {
+    lock = call_combiner->tsan_lock;
+    TSAN_ANNOTATE_RWLOCK_ACQUIRED(lock.get(), true);
+    lock->taken.store(true, std::memory_order_relaxed);
+  }
+  GRPC_CLOSURE_RUN(call_combiner->original_closure, GRPC_ERROR_REF(error));
```

Comment:
It looks like what's happening here is that we're arranging to hold the lock for the duration of the closure's executing *in this thread only* -- in other words, we give up the lock when the callback returns, so if the callback is still holding the call combiner while doing some asynchronous work (which will be picked up later in a different thread), the lock is not held while that asynchronous work is done.

Is that correct?  If so, doesn't that mean that there could still be problems if work done is asynchronously in two different callbacks?  Or would the fact that we're acquiring this when each callback starts avoid that problem somehow?

Sorry if these are dumb questions -- I always find it hard reasoning about how tsan works for some reason.

Reply:
That's a good question. If the callback is scheduled via the call_combiner's start/finish, we will get the locking annotations, but if they go via other mechanism and happen to access the fields in parallel to other threads, these annotations wouldn't help. 

Running all the TSAN tests, I don't think we have a case where TSAN is complaining about. If later we hit an error like that, we would need to add an extra API for such callbacks (either explicit lock/unlock or something like call_combiner_sched_closure).

Reply:
TSAN nicely detected an error in my previous implementation, which was using load/store instead of compare_exchange. :-) It couldn't prove I'm doing it safely.

Reply:
Okay, sounds like we have a potential hole here, then.  I'm fine with not worrying about this until we actually have a problem with it, but how about adding a TODO comment explaining when this might happen, as a clue to someone trying to debug this kind of problem in the future?